cmd/agent/app/reporter/grpc/reporter.go

```diff
@@ -55,14 +53,14 @@ func (r *Reporter) EmitZipkinBatch(zSpans []*zipkincore.Span) error {
 	if err != nil {
 		return err
 	}
-	return r.send(trace.Spans)
-}
-
-func (r *Reporter) send(spans []*model.Span) error {
 	var process model.Process
-	if len(spans) > 0 {
-		process = *spans[0].Process
+	if len(trace.GetSpans()) > 0 {
+		process = *trace.GetSpans()[0].Process
 	}
+	return r.send(trace.Spans, process)
+}
+
+func (r *Reporter) send(spans []*model.Span, process model.Process) error {
```

Comment:
Maybe we should pass process=nil for zipkin path? I don't recall how collector handles batches, it should not apply batch-level Process to each span if spans already have a Process attached. This would be more compatible with, say, Istio use case where one client can emit spans for different processes.

Once we do that, we can kill this zipkin code path in the processor API and pull it higher in the chain, i.e. since you're already calling toDomain in the grpc package (but not in tchannel package), if we confirm that we don't cause a data loss we can call toDomain before invoking the processor.

NB: we may still need to keep the zipkin tchannel endpoint in the collector because some very old clients may be using direct tchannel connections to the collector, although I find it hard to believe since the agent was implemented long before we open sourced Jaeger backend (like 1yr before), so nobody outside of Uber should be using those old clients. If so, we can remove that endpoint and re-keep it in our internal build only.

Reply:
>  I don't recall how collector handles batches

```go
// PostSpans implements gRPC CollectorService.
func (g *GRPCHandler) PostSpans(ctx context.Context, r *api_v2.PostSpansRequest) (*api_v2.PostSpansResponse, error) {
	for _, span := range r.GetBatch().Spans {
		if span.GetProcess() == nil {
			span.Process = &r.Batch.Process
		}
	}
	_, err := g.spanProcessor.ProcessSpans(r.GetBatch().Spans, JaegerFormatType)
	if err != nil {
		g.logger.Error("cannot process spans", zap.Error(err))
		return nil, err
	}
	return &api_v2.PostSpansResponse{}, nil
}
```

It sets the process only when it is missing.

Another issue is that this zipkin path skips zipkin sanitizers which are used in collector handler.

Reply:
> Once we do that, we can kill this zipkin code path in the processor API and pull it higher in the chain, i.e. since you're already calling toDomain in the grpc package (but not in tchannel package), if we confirm that we don't cause a data loss we can call toDomain before invoking the processor.

I don't get you here. Once we do what pass the nill process? I have just added sanitizers directly to the agent.


> so nobody outside of Uber should be using those old clients. If so, we can remove that endpoint and re-keep it in our internal build only.
tracked in  https://github.com/jaegertracing/jaeger/issues/1186 for

Reply:
Right now we have this flow:
```
udp server jaeger.thrift -> reporter::SubmitBatch
udp server zipking.thrift -> reporter::SubmitZipkinSpans

tchannel reporter::SubmitBatch -> pass through to collector
tchannel reporter::SubmitZipkinSpans -> pass through to collector

grpc reporter::SubmitBatch -> convert to domain and send to collector
grpc reporter::SubmitZipkinSpans -> convert to jaeger.thrift, then to domain and send to collector
```
What I am proposing (could be done in a different PR), is to eliminate `reporter::SubmitZipkinSpans` completely and have this:
```
udp server jaeger.thrift -> reporter::SubmitBatch
    tchannel reporter::SubmitBatch -> pass through to collector
    grpc reporter::SubmitBatch -> convert to domain and send to collector
udp server zipking.thrift -> convert to jaeger.thrift -> reporter::SubmitBatch
```
This way the Reporter only has a single method for Jaeger spans (in Thrift, for now).

Once we remove tchannel completely, we can push thrift->domain conversion into UDP servers as well, and Reporter interface effectively becomes == Collector service from proto IDL.

Reply:
I see, it sounds reasonable to me. I would prefer a separate PR to do it as it might be bigger than I think right now.

Reply:
I have created to track it separately https://github.com/jaegertracing/jaeger/issues/1193